File: src/configcontainer.cpp
#include "configcontainer.h"

#include <algorithm>
#include <cassert>
#include <iostream>
#include <pwd.h>
#include <sstream>
#include <sys/types.h>

#include "config.h"
#include "configparser.h"
#include "confighandlerexception.h"
#include "logger.h"
#include "strprintf.h"
#include "utils.h"

namespace newsboat {

const std::string ConfigContainer::PARTIAL_FILE_SUFFIX = "part";

ConfigContainer::ConfigContainer()
	: rs_object(newsboat::configcontainer::bridged::create())
{
}

ConfigContainer::~ConfigContainer() = default;

void ConfigContainer::register_commands(ConfigParser& cfgparser)
{
	// Empty string in get_suggestions() returns all available keys which we need to register
	const auto keys = newsboat::configcontainer::bridged::get_suggestions(*rs_object, "");
	for (const auto& key : keys) {
		cfgparser.register_handler(std::string(key), *this);
	}
}

void ConfigContainer::handle_action(const std::string& action,
	const std::vector<std::string>& params)
{
	rust::Vec<rust::String> rs_params;
	for (const auto& p : params) {
		rs_params.push_back(p);
	}

	const auto result = newsboat::configcontainer::bridged::handle_action(*rs_object, action,
			rs_params);

	// Map status codes from Rust (ConfigHandlerStatus) to C++ ActionHandlerStatus
	switch (result.status) {
	case 0: // Success
		break;
	case 1: // InvalidCommand
		throw ConfigHandlerException(ActionHandlerStatus::INVALID_COMMAND);
	case 2: // TooFewParams
		throw ConfigHandlerException(ActionHandlerStatus::TOO_FEW_PARAMS);
	case 3: // TooManyParams
		throw ConfigHandlerException(ActionHandlerStatus::TOO_MANY_PARAMS);
	case 4: // InvalidParams
		throw ConfigHandlerException(std::string(result.msg));
	default:
		throw ConfigHandlerException(std::string("Unknown error from Rust"));
	}
}

std::string ConfigContainer::get_configvalue(const std::string& key) const
{
	return std::string(newsboat::configcontainer::bridged::get_configvalue(*rs_object, key));
}

int ConfigContainer::get_configvalue_as_int(const std::string& key) const
{
	return newsboat::configcontainer::bridged::get_configvalue_as_int(*rs_object, key);
}

Filepath ConfigContainer::get_configvalue_as_filepath(const std::string& key) const
{
	return Filepath(newsboat::configcontainer::bridged::get_configvalue_as_filepath(*rs_object,
				key));
}

bool ConfigContainer::get_configvalue_as_bool(const std::string& key) const
{
	return newsboat::configcontainer::bridged::get_configvalue_as_bool(*rs_object, key);
}

nonstd::expected<void, std::string> ConfigContainer::set_configvalue(
	const std::string& key,
	const std::string& value)
{
	rust::String error_message;
	if (newsboat::configcontainer::bridged::set_configvalue(*rs_object, key, value,
			error_message)) {
		return {};
	} else {
		return nonstd::make_unexpected(std::string(error_message));
	}
}

void ConfigContainer::reset_to_default(const std::string& key)
{
	newsboat::configcontainer::bridged::reset_to_default(*rs_object, key);
}

void ConfigContainer::toggle(const std::string& key)
{
	newsboat::configcontainer::bridged::toggle(*rs_object, key);
}

void ConfigContainer::dump_config(std::vector<std::string>& config_output) const
{
	const auto lines = newsboat::configcontainer::bridged::dump_config(*rs_object);
	for (const auto& line : lines) {
		config_output.push_back(std::string(line));
	}
}

std::vector<std::string> ConfigContainer::get_suggestions(
	const std::string& fragment) const
{
	const auto rs_suggs = newsboat::configcontainer::bridged::get_suggestions(*rs_object,
			fragment);
	std::vector<std::string> result;
	for (const auto& s : rs_suggs) {
		result.push_back(std::string(s));
	}
	return result;
}

FeedSortStrategy ConfigContainer::get_feed_sort_strategy() const
{
	FeedSortStrategy ss;
	const auto values = newsboat::configcontainer::bridged::get_feed_sort_strategy_values(
			*rs_object);
	assert(values.size() >= 2);
	ss.sm = static_cast<FeedSortMethod>(values[0]);
	ss.sd = static_cast<SortDirection>(values[1]);
	return ss;
}

ArticleSortStrategy ConfigContainer::get_article_sort_strategy() const
{
	ArticleSortStrategy ss;
	const auto values = newsboat::configcontainer::bridged::get_article_sort_strategy_values(
			*rs_object);
	assert(values.size() >= 2);
	ss.sm = static_cast<ArtSortMethod>(values[0]);
	ss.sd = static_cast<SortDirection>(values[1]);
	return ss;
}

} // namespace newsboat

File: rust/libnewsboat-ffi/src/configcontainer.rs
use crate::filepath::PathBuf;
use cxx::{ExternType, type_id};
use libnewsboat::configcontainer;
use std::pin::Pin;

pub struct ConfigContainer(pub configcontainer::ConfigContainer);

unsafe impl ExternType for ConfigContainer {
    type Id = type_id!("newsboat::configcontainer::bridged::ConfigContainer");
    type Kind = cxx::kind::Opaque;
}

#[cxx::bridge(namespace = "newsboat::configcontainer::bridged")]
mod bridged {
    #[namespace = "newsboat::filepath::bridged"]
    extern "C++" {
        include!("libnewsboat-ffi/src/filepath.rs.h");
        type PathBuf = crate::filepath::PathBuf;
    }

    struct CfgHandlerResult {
        status: u8,
        msg: String,
    }

    extern "Rust" {
        type ConfigContainer;

        fn create() -> Box<ConfigContainer>;

        fn handle_action(
            cc: &ConfigContainer,
            action: &str,
            params: Vec<String>,
        ) -> CfgHandlerResult;

        fn get_configvalue(cc: &ConfigContainer, key: &str) -> String;
        fn get_configvalue_as_int(cc: &ConfigContainer, key: &str) -> i32;
        fn get_configvalue_as_bool(cc: &ConfigContainer, key: &str) -> bool;
        fn get_configvalue_as_filepath(
            cc: &ConfigContainer,
            key: &str,
            path: Pin<&mut PathBuf>,
        ) -> bool;

        fn set_configvalue(
            cc: &ConfigContainer,
            key: &str,
            value: &str,
            error_msg: &mut String,
        ) -> bool;
        fn reset_to_default(cc: &ConfigContainer, key: &str);
        fn toggle(cc: &ConfigContainer, key: &str);
        fn dump_config(cc: &ConfigContainer) -> Vec<String>;
        fn get_suggestions(cc: &ConfigContainer, fragment: &str) -> Vec<String>;

        fn get_feed_sort_strategy_values(cc: &ConfigContainer) -> Vec<i32>;
        fn get_article_sort_strategy_values(cc: &ConfigContainer) -> Vec<i32>;
    }
}

fn create() -> Box<ConfigContainer> {
    Box::new(ConfigContainer(configcontainer::ConfigContainer::new()))
}

fn handle_action(
    cc: &ConfigContainer,
    action: &str,
    params: Vec<String>,
) -> bridged::CfgHandlerResult {
    match cc.0.handle_action(action, &params) {
        configcontainer::ConfigHandlerStatus::Success => bridged::CfgHandlerResult {
            status: 0,
            msg: String::new(),
        },
        configcontainer::ConfigHandlerStatus::InvalidCommand => bridged::CfgHandlerResult {
            status: 1,
            msg: String::new(),
        },
        configcontainer::ConfigHandlerStatus::TooFewParams => bridged::CfgHandlerResult {
            status: 2,
            msg: String::new(),
        },
        configcontainer::ConfigHandlerStatus::TooManyParams => bridged::CfgHandlerResult {
            status: 3,
            msg: String::new(),
        },
        configcontainer::ConfigHandlerStatus::InvalidParams(m) => {
            bridged::CfgHandlerResult { status: 4, msg: m }
        }
    }
}

fn get_configvalue(cc: &ConfigContainer, key: &str) -> String {
    cc.0.get_configvalue(key)
}

fn get_configvalue_as_int(cc: &ConfigContainer, key: &str) -> i32 {
    cc.0.get_configvalue_as_int(key)
}

fn get_configvalue_as_bool(cc: &ConfigContainer, key: &str) -> bool {
    cc.0.get_configvalue_as_bool(key)
}

fn get_configvalue_as_filepath(
    cc: &ConfigContainer,
    key: &str,
    mut path: Pin<&mut PathBuf>,
) -> bool {
    cc.0.get_configvalue_as_filepath(key, &mut path.get_mut().0)
}

fn set_configvalue(cc: &ConfigContainer, key: &str, value: &str, error_msg: &mut String) -> bool {
    match cc.0.set_configvalue(key, value) {
        Ok(_) => true,
        Err(e) => {
            *error_msg = e;
            false
        }
    }
}

fn reset_to_default(cc: &ConfigContainer, key: &str) {
    cc.0.reset_to_default(key);
}

fn toggle(cc: &ConfigContainer, key: &str) {
    cc.0.toggle(key);
}

fn dump_config(cc: &ConfigContainer) -> Vec<String> {
    cc.0.dump_config()
}

fn get_suggestions(cc: &ConfigContainer, fragment: &str) -> Vec<String> {
    cc.0.get_suggestions(fragment)
}

fn get_feed_sort_strategy_values(cc: &ConfigContainer) -> Vec<i32> {
    let (method, direction) = cc.0.get_feed_sort_strategy();

    let m = match method {
        configcontainer::FeedSortMethod::None => 0,
        configcontainer::FeedSortMethod::FirstTag => 1,
        configcontainer::FeedSortMethod::Title => 2,
        configcontainer::FeedSortMethod::ArticleCount => 3,
        configcontainer::FeedSortMethod::UnreadArticleCount => 4,
        configcontainer::FeedSortMethod::LastUpdated => 5,
        configcontainer::FeedSortMethod::LatestUnread => 6,
    };

    let d = match direction {
        configcontainer::SortDirection::Asc => 0,
        configcontainer::SortDirection::Desc => 1,
    };

    vec![m, d]
}

fn get_article_sort_strategy_values(cc: &ConfigContainer) -> Vec<i32> {
    let (method, direction) = cc.0.get_article_sort_strategy();

    let m = match method {
        configcontainer::ArtSortMethod::Title => 0,
        configcontainer::ArtSortMethod::Flags => 1,
        configcontainer::ArtSortMethod::Author => 2,
        configcontainer::ArtSortMethod::Link => 3,
        configcontainer::ArtSortMethod::Guid => 4,
        configcontainer::ArtSortMethod::Date => 5,
        configcontainer::ArtSortMethod::Random => 6,
    };

    let d = match direction {
        configcontainer::SortDirection::Asc => 0,
        configcontainer::SortDirection::Desc => 1,
    };

    vec![m, d]
}
